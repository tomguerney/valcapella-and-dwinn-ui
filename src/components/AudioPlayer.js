import React from 'react'
import Button from '@material-ui/core/Button'
import AudioSlider from './AudioSlider'
import PlayStatusIcon from './PlayStatusIcon'
import { Grid } from '@material-ui/core'

export default class AudioPlayer extends React.Component {
  state = {
    isPlaying: false,
    isLoading: false,
    duration: 0,
    currentTime: 0,
  }
  audio = React.createRef()
  componentDidUpdate = (prevProps) => {
    if (prevProps.audioHidden !== this.props.audioHidden) {
      this.pause()
    }
  }
  onTogglePlay = () => {
    this.state.isPlaying ? this.pause() : this.play()
  }
  play = () => {
    if (!this.audio.current.src || this.audio.current.src.length === 0) {
      this.audio.current.src = this.props.audioSource
    }
    this.setState(() => ({ isPlaying: true, isLoading: true }))
    this.audio.current.play()
  }
  pause = () => {
    this.setState(() => ({ isPlaying: false }))
    this.audio.current.pause()
  }
  onUpdateTime = () => {
    this.setState(() => ({ currentTime: this.audio.current.currentTime }))
  }
  onDurationChange = () => {
    this.setState(() => ({ duration: this.audio.current.duration }))
  }
  onWaiting = () => {
    this.setState(() => ({ isLoading: true }))
  }
  onPlaying = () => {
    this.setState(() => ({ isLoading: false, isPlaying: true }))
  }
  onEnded = () => {
    this.setState(() => ({ currentTime: 0, isPlaying: false }))
  }
  onSliderChange = (event, value) => {
    this.setState(() => ({ currentTime: value }))
    this.audio.current.currentTime = value
  }
  render() {
    return (
      <React.Fragment>
        <audio
          ref={this.audio}
          onTimeUpdate={this.onUpdateTime}
          onDurationChange={this.onDurationChange}
          onWaiting={this.onWaiting}
          onPlaying={this.onPlaying}
          onEnded={this.onEnded}
        />
        <table style={{ paddingBottom: 15, width: '100%' }}>
          <tbody>
            <tr>
              <td
                style={{
                  width: '80px',
                  maxWidth: '80px',
                  mindWidth: '80px',
                  paddingLeft: 15,
                  paddingRight: 15,
                }}>
                <Button
                  variant='outlined'
                  style={{ minHeight: 45, minWidth: 60, backgroundColor: '#C8DEF4' }}
                  onClick={this.onTogglePlay}>
                  <PlayStatusIcon isPlaying={this.state.isPlaying} isLoading={this.state.isLoading} />
                </Button>
              </td>
              <td style={{ paddingLeft: 15, paddingRight: 15 }}>
                <Grid container justify='center'>
                  <AudioSlider
                    max={this.state.duration}
                    value={this.state.currentTime}
                    onChange={this.onSliderChange}
                  />
                </Grid>
              </td>
            </tr>
          </tbody>
        </table>
      </React.Fragment>
    )
  }
}
