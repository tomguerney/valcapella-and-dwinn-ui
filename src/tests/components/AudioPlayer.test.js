import React from 'react'
import { shallow } from 'enzyme'
import AudioPlayer from '../../components/AudioPlayer'

let wrapper, instance, mockAudioSource, audioMock

beforeEach(() => {
  mockAudioSource = 'mock_audio_source.mp3'
  audioMock = {
    src: '',
    play: jest.fn(),
    pause: jest.fn(),
    currentTime: 0,
  }
  wrapper = shallow(<AudioPlayer audioHidden={true} audioSource={mockAudioSource} />)
  instance = wrapper.instance()
  instance.audio.current = audioMock
})

test('should render AudioPlayer correctly', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should not call pause when audioHidden is true and updated to true ', () => {
  instance.pause = jest.fn()
  wrapper.setProps({ audioHidden: true })
  expect(instance.pause).not.toHaveBeenCalled()
})

test('should call pause when audioHidden is false and updated to true ', () => {
  const lclWrapper = shallow(<AudioPlayer audioHidden={false} />)
  const lclInstance = lclWrapper.instance()
  lclInstance.pause = jest.fn()
  lclWrapper.setProps({ audioHidden: true })
  expect(lclInstance.pause).toHaveBeenCalled()
})

test('should call play when isPlaying is false and button is clicked', () => {
  instance.play = jest.fn()
  wrapper.find('WithStyles(ForwardRef(Button))').simulate('click')
  expect(instance.play).toHaveBeenCalled()
})

test('should call pause when isPlaying is true and button is clicked', () => {
  wrapper.setState({ isPlaying: true })
  instance.pause = jest.fn()
  wrapper.find('WithStyles(ForwardRef(Button))').simulate('click')
  expect(instance.pause).toHaveBeenCalled()
})

test('should handle play with unset audio source', () => {
  instance.play()
  expect(instance.audio.current.src).toBe(mockAudioSource)
  expect(audioMock.play).toHaveBeenCalled()
  expect(wrapper.state()).toMatchObject({ isPlaying: true, isLoading: true })
})

test('should handle play with set audio source', () => {
  const existingAudioSource = 'set_audio_src.mp3'
  instance.audio.current.src = existingAudioSource
  instance.play()
  expect(instance.audio.current.src).toBe(existingAudioSource)
  expect(audioMock.play).toHaveBeenCalled()
  expect(wrapper.state()).toMatchObject({ isPlaying: true, isLoading: true })
})

test('should handle pause correctly', () => {
  wrapper.setState({ isPlaying: true })
  instance.pause()
  expect(audioMock.pause).toHaveBeenCalled()
  expect(wrapper.state('isPlaying')).toBe(false)
})

test('should handle onTimeUpdate correctly', () => {
  const currentTime = 101
  audioMock.currentTime = currentTime
  wrapper.find('audio').prop('onTimeUpdate')()
  expect(wrapper.state('currentTime')).toBe(currentTime)
})

test('should handle onDurationChange correctly', () => {
  const duration = 1001
  audioMock.duration = duration
  wrapper.find('audio').prop('onDurationChange')()
  expect(wrapper.state('duration')).toBe(duration)
})

test('should hanel onWaiting correctly', () => {
  wrapper.find('audio').prop('onWaiting')()
  expect(wrapper.state('isLoading')).toBe(true)
})

test('should handle onPlaying correctly', () => {
  wrapper.find('audio').prop('onPlaying')()
  expect(wrapper.state()).toMatchObject({ isPlaying: true, isLoading: false })
})

test('should handle onEnded correctly', () => {
  wrapper.setState({ currentTime: 99, isPlaying: true })
  wrapper.find('audio').prop('onEnded')()
  expect(wrapper.state()).toMatchObject({ currentTime: 0, isPlaying: false })
})

test('should handle onSliderChange correctly', () => {
  const newTime = 303
  wrapper.find('WithStyles(WithStyles(ForwardRef(Slider)))').prop('onChange')({}, newTime)
  expect(wrapper.state('currentTime')).toBe(newTime)
  expect(audioMock.currentTime).toBe(newTime)
})
