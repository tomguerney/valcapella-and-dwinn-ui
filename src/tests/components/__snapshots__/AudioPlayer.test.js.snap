// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AudioPlayer correctly 1`] = `
<Fragment>
  <audio
    onDurationChange={[Function]}
    onEnded={[Function]}
    onPlaying={[Function]}
    onTimeUpdate={[Function]}
    onWaiting={[Function]}
  />
  <table
    style={
      Object {
        "paddingBottom": 15,
        "width": "100%",
      }
    }
  >
    <tbody>
      <tr>
        <td
          style={
            Object {
              "maxWidth": "80px",
              "mindWidth": "80px",
              "paddingLeft": 15,
              "paddingRight": 15,
              "width": "80px",
            }
          }
        >
          <WithStyles(ForwardRef(Button))
            onClick={[Function]}
            style={
              Object {
                "backgroundColor": "#C8DEF4",
                "minHeight": 45,
                "minWidth": 60,
              }
            }
            variant="outlined"
          >
            <PlayStatusIcon
              isLoading={false}
              isPlaying={false}
            />
          </WithStyles(ForwardRef(Button))>
        </td>
        <td
          style={
            Object {
              "paddingLeft": 15,
              "paddingRight": 15,
            }
          }
        >
          <WithStyles(ForwardRef(Grid))
            container={true}
            justify="center"
          >
            <WithStyles(WithStyles(ForwardRef(Slider)))
              max={0}
              onChange={[Function]}
              value={0}
            />
          </WithStyles(ForwardRef(Grid))>
        </td>
      </tr>
    </tbody>
  </table>
</Fragment>
`;
